src/nanoemoji/util.py

```diff
@@ -85,30 +87,120 @@ def file_printer(filename):
+    for tag in font.keys():
+        table = font[tag]
         # cmap is exceptional in that it always loads subtables lazily upon getting
         # their attributes, no matter the value of TTFont.lazy option.
         # TODO: remove this hack once fixed in fonttools upstream
         if tag == "cmap":
             _ = [st.cmap for st in table.tables]
-        ttfont[tag] = table
 
+    _assert_all_keys_loaded(font)
+
+    # post has a cool pattern of raising errors if it has no glyphOrder attribute
+    # font isn't all that useful w/o glyph order so lets confirm we're ok on that front
+    # TODO contemplate whether this is simply a FontTools bug
+    font.getGlyphOrder()
+
+    return font
+
+
+class OtDataPathEntry(NamedTuple):
+    value: Any
+    attr_name: str  # attr of parent to get value
+
+
+OtDataPath = Tuple[OtDataPathEntry, ...]
+
+
+def traverse_ot_data(root, callback: Callable[[OtDataPath], None]) -> Iterable[Any]:
+    visited = set()
+    frontier: List[OtDataPath] = [(OtDataPathEntry(root, ""),)]
+    while frontier:
+        # path is (value, attr_name) tuples. attr_name is attr of parent to get value
+        path = frontier.pop(0)
+        current = path[-1].value
+        visited.add(id(current))
 
-def reorder_glyphs(ttfont: ttLib.TTFont, glyph_order: List[str]):
+        callback(path)
+
+        converters = getattr(current, "converters", [])
+        for converter in converters:
+            is_simple = isinstance(converter, otConverters.SimpleValue)
+            print("  ", getattr(converter, "name", "?nameless?"), converter, is_simple)
+
+            attr_name = getattr(converter, "name", None)
+            if attr_name:
+                raw_value = getattr(current, attr_name, None)
+            else:
+                print("  ", "^has no name")
+                raw_value = None
+
+            if not isinstance(raw_value, list):
+                raw_value = [raw_value]
+
+            for value in raw_value:
+                if (
+                    value is not None
+                    and hasattr(value, "converters")
+                    and id(value) not in visited
+                ):
+                    frontier.append(path + (OtDataPathEntry(value, attr_name),))
+
+
+def reorder_glyphs(font: ttLib.TTFont, glyph_order: List[str]):
     # Changing the order of glyphs in a TTFont requires that all tables that use
     # glyph indexes have been fully decompiled (loaded with lazy=False).
     # Cf. https://github.com/fonttools/fonttools/issues/2060
-    ensure_ttfont_fully_decompiled(ttfont)
-    ttfont.setGlyphOrder(glyph_order)
+
+    _assert_all_keys_loaded(font)
+
+    font.setGlyphOrder(glyph_order)
     # glyf table is special and needs its own glyphOrder...
-    assert ttfont.isLoaded("glyf")
-    ttfont["glyf"].glyphOrder = glyph_order
+    font["glyf"].glyphOrder = glyph_order
+
+    # make sure reverse glyph order rebuilds too
+    font.getReverseGlyphMap(rebuild=True)
+
+    # TEMPORARY HOPEFULLY
```

Comment:
well not temporary, it's just the way otData-tables work in fonttools; there are parallel lists (various Coverage and linked arrays) that need to be kept in the same order; changing the glyph order requires one to change the order of those tables as well, as we are doing below. It's more likely that fonttools grows an additional API to reorder glyphs which transparently does all this work for us, than Coverage and linked lists updating automatically on compile as the glyphOrder changes. So I think that warning is not gonna go away. The attempt at self-repairing is some obscure hack that was put in place for some SIL fonts
https://github.com/fonttools/fonttools/blob/71a5bf1b512e51044fe2af31d86f2e73bd3a3eaa/Lib/fontTools/ttLib/tables/otTables.py#L527-L529
I'm not even sure it ever worked tbh

Reply:
Oops, that's a stale comment from before you clarified the need to sort things in ||